packages/verified-fetch/src/utils/byte-range-context.ts

```diff
@@ -0,0 +1,287 @@
+  private readonly _contentRangeHeaderValue: string | undefined
+  private _body: SupportedBodyTypes | null = null
+  private readonly _rangeRequestHeader: string | undefined
+  private readonly log: Logger
+  private _isValidRangeRequest: boolean | null = null
+  private readonly requestRangeStart: number | null
+  private readonly requestRangeEnd: number | null
+  private byteStart: number | undefined
+  private byteEnd: number | undefined
+  private byteSize: number | undefined
+
+  constructor (logger: ComponentLogger, private readonly headers?: HeadersInit) {
+    this.log = logger.forComponent('helia:verified-fetch:byte-range-context')
+    this._rangeRequestHeader = getHeader(this.headers, 'Range')
+    if (this._rangeRequestHeader != null) {
+      this.log.trace('range request detected')
+      this._isRangeRequest = true
+      try {
+        const { start, end } = getByteRangeFromHeader(this._rangeRequestHeader)
+        this.requestRangeStart = start != null ? parseInt(start) : null
+        this.requestRangeEnd = end != null ? parseInt(end) : null
+      } catch (e) {
+        this.log.error('error parsing range request header: %o', e)
+        this.isValidRangeRequest = false
+        this.requestRangeStart = null
+        this.requestRangeEnd = null
+      }
+
+      this.setOffsetDetails()
+    } else {
+      this.log.trace('no range request detected')
+      this._isRangeRequest = false
+      this.requestRangeStart = null
+      this.requestRangeEnd = null
+    }
+  }
+
+  public setBody (body: SupportedBodyTypes): void {
+    this._body = body
+    // if fileSize was already set, don't recalculate it
+    this.fileSize = this.fileSize ?? getBodySizeSync(body)
+
+    this.log.trace('set request body with fileSize %o', this._fileSize)
+  }
+
+  public getBody (): SupportedBodyTypes {
+    const body = this._body
+    if (body == null) {
+      this.log.trace('body is null')
+      return body
+    }
+    if (!this.isRangeRequest || !this.isValidRangeRequest) {
+      this.log.trace('returning body unmodified for non-range, or invalid range, request')
+      return body
+    }
+    const byteStart = this.byteStart
+    const byteEnd = this.byteEnd
+    const byteSize = this.byteSize
+    if (byteStart != null || byteEnd != null) {
+      this.log.trace('returning body with byteStart=%o, byteEnd=%o, byteSize=%o', byteStart, byteEnd, byteSize)
+      if (body instanceof ReadableStream) {
+        // stream should already be spliced by `unixfs.cat`
+        return body
+      }
+      return this.getSlicedBody(body)
+    }
+
+    // we should not reach this point, but return body untouched.
+    this.log.error('returning unmofified body for valid range request')
+    return body
+  }
+
+  private getSlicedBody <T extends Uint8Array | ArrayBuffer | Blob | string>(body: T): T {
+    if (this.isPrefixLengthRequest) {
+      this.log.trace('sliced body with byteStart %o', this.byteStart)
+      return body.slice(this.offset) as T
```

Comment:
All this casting removes any type safety we might have.  Would it be better to convert any passed body types to just one type instead of having to deal with all of: `Uint8Array | ArrayBuffer | Blob | string`?

Convert everything to `Blob` or something?